libkatana_python_native/include/katana/python/Conventions.h

```diff
@@ -0,0 +1,95 @@
+#ifndef KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_CONVENTIONS_H_
+#define KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_CONVENTIONS_H_
+
+#include <type_traits>
+
+#include <pybind11/pybind11.h>
+
+namespace katana {
+
+namespace detail {
+
+/// DefRepr will def `__repr__` based on `ToString` if it is available.
+template <typename T, typename Enable = void>
+struct DefRepr {
+  void operator()(pybind11::class_<T>& cls [[maybe_unused]]) {}
+};
+
+template <typename T>
+struct DefRepr<T, std::void_t<decltype(std::declval<T>().ToString())>> {
+  void operator()(pybind11::class_<T>& cls) {
+    cls.def("__repr__", &T::ToString);
+  }
+};
+
+/// DefEqualsEquals will def `__eq__` based on `operator==` if it is available.
+template <typename T, typename Enable = void>
+struct DefEqualsEquals {
+  void operator()(pybind11::class_<T>& cls [[maybe_unused]]) {}
+};
+
+template <typename T>
+struct DefEqualsEquals<
+    T, std::void_t<decltype(std::declval<T>() == std::declval<T>())>> {
+  void operator()(pybind11::class_<T>& cls) {
+    cls.def("__eq__", &T::operator==);
+  }
+};
+
+/// DefEqualsEquals will def `__eq__` based on `Equals` or `operator==` if one
+/// is available (`Equals` is preferred).
+template <typename T, typename Enable = void>
+struct DefEquals {
+  void operator()(pybind11::class_<T>& cls) {
+    // If Equals doesn't exist then check for ==.
+    DefEqualsEquals<T>{}(cls);
+  }
+};
+
+template <typename T>
+struct DefEquals<
+    T, std::void_t<decltype(std::declval<T>().Equals(std::declval<T>()))>> {
+  void operator()(pybind11::class_<T>& cls) { cls.def("__eq__", &T::Equals); }
+};
+
+/// DefCopy defs `__copy__` and `copy` based on the copy constructor if it is
+/// available.
+template <typename T, typename Enable = void>
+struct DefCopy {
```

Comment:
I understand the python to c++ mapping but for each, what are these maybe_unused portions?

Reply:
Those are the "default" cases which get used if the type involved *does not* have the methods or operators I'm checking for. This is used to allow `DefConventions` to be called on any type and it will bind whatever conventions are actually present on the type and ignore any that are not.